owb.c

```diff
@@ -443,64 +135,85 @@ void owb_free(OneWireBus ** bus)
     }
 }
 
-void owb_init(OneWireBus * bus, int gpio)
+owb_status owb_init(OneWireBus * bus, int gpio)
 {
-    if (bus != NULL)
-    {
-        bus->gpio = gpio;
-        bus->timing = &_StandardTiming;
-        bus->init = true;
+    owb_status status;
 
-        // platform specific:
-        gpio_pad_select_gpio(bus->gpio);
-    }
-    else
-    {
-        ESP_LOGE(TAG, "bus is NULL");
-    }
+    // default to the gpio driver for now
+    bus->driver = owb_gpio_get_driver();
+
+    status = bus->driver->init(bus, gpio);
+
+    return status;
 }
 
-void owb_use_crc(OneWireBus * bus, bool use_crc)
+owb_status owb_use_crc(OneWireBus * bus, bool use_crc)
 {
+    owb_status status;
+
     if (_is_init(bus))
     {
         bus->use_crc = use_crc;
         ESP_LOGD(TAG, "use_crc %d", bus->use_crc);
+
+        status = OWB_STATUS_OK;
+    } else
+    {
+        status = OWB_STATUS_NOT_INITIALIZED;
     }
+
+    return status;
 }
 
-OneWireBus_ROMCode owb_read_rom(const OneWireBus * bus)
+owb_status owb_read_rom(const OneWireBus * bus, OneWireBus_ROMCode *rom_code)
 {
-    OneWireBus_ROMCode rom_code = {0};
+    owb_status status = OWB_STATUS_ERR;
+
+    memset(rom_code, 0, sizeof(OneWireBus_ROMCode));
+
     if (_is_init(bus))
     {
-        if (_reset(bus))
+        bool is_present;
+        bus->driver->reset(bus, &is_present);
+        if (is_present)
         {
-            _write_byte(bus, OWB_ROM_READ);
-            _read_block(bus, rom_code.bytes, sizeof(rom_code));
+            uint8_t value = OWB_ROM_READ;
+            bus->driver->write_bytes(bus, &value, sizeof(value));
+            bus->driver->read_bytes(bus, rom_code->bytes, sizeof(OneWireBus_ROMCode));
 
             if (bus->use_crc)
             {
-                if (owb_crc8_bytes(0, rom_code.bytes, sizeof(rom_code)) != 0)
+                if (owb_crc8_bytes(0, rom_code->bytes, sizeof(OneWireBus_ROMCode)) != 0)
```

Comment:
In general, where return value is now a pass-by-pointer parameter, I think there should be a null check before dereference. I would usually put this very early in the function with an appropriate error code or trace message. I think there are a few cases of this now.

Reply:
@DavidAntliff  let me check to see what happens if we dereference close to zero. I do prefer c++ for references, they tend to reduce these kinds of issues.

I think I figured out how to do it dynamically and with all of the requirements. I should have something tomorrow.

Reply:
@DavidAntliff and on the common vs. different I think it will become more obvious when the rmt is merged in. I'll go back and fix the original driver changes as more stuff drops out. I'm hoping to have the rmt stuff up and running tomorrow or the day after. It's been a fun excursion but I've got a lot of stuff in the queue!

Reply:
Yes, C++ references help a lot to avoid the issues around null pointers. Unfortunately we don't have that luxury here. The check is more to catch errors when a user accidentally passes NULL via the API - I do try to make all my API functions "uncrashable" (at least insofar as simple data types).

The check can be very simple, e.g. https://github.com/DavidAntliff/esp32-tsl2561/blob/master/tsl2561.c#L324

Reply:
@DavidAntliff ok. I can fit that into an earlier commit.